### PR TITLE
abinit: make libxml2 really optional, add optimization-flavor variant, fix build with ifort for atompaw

### PR DESCRIPTION
**Pull request by dev-zero**

### CODE REVIEW

**Comment by dev-zero** [comment]
@glennpj since you seem to have updated it recently: I am still unable to build with `+openmp` when using an Intel compiler. So far it seems Abinit is using the CDFT interface of MKL. If that supports OpenMP transparently, we could simply restructure the conditional to:

```
if '^mkl' in spec:
  ...
elif '^fftw' in spec:
    if '+openmp' in spec:
        ...
    else:
        ...
```

If not, an option would be to switch to MKLs FFTW3 API instead?

**Comment by glennpj** [comment]
> @glennpj since you seem to have updated it recently: I am still unable to build with `+openmp` when using an Intel compiler. So far it seems Abinit is using the CDFT interface of MKL. If that supports OpenMP transparently, we could simply restructure the conditional to:
> 
> ```
> if '^mkl' in spec:
>   ...
> elif '^fftw' in spec:
>     if '+openmp' in spec:
>         ...
>     else:
>         ...
> ```
> 
> If not, an option would be to switch to MKLs FFTW3 API instead?

We should disable `+openmp` with Intel FFT, ie., `+openmp` requires `FFTW3`.

**Review by glennpj — DISMISSED** [review]
This looks good to me. Not sure if you want to address the MKL FFTW issue in this PR or not.

**Comment by dev-zero** [comment]
@glennpj thanks for the review. I've looked into the code and from what I've seen it seems that Abinit is checking for threads in the CDFT interface at runtime. Based on that I built a version with Abinit with the following logic:

```python
            if '^mkl' in spec:
                fftflavor = 'dfti'
                options.extend([
                    'FFT_CPPFLAGS={0}'.format(spec['fftw-api'].headers.cpp_flags),
                    'FFT_LIBs={0}'.format(spec['fftw-api'].libs.ld_flags),
                ])
            elif '^fftw' in spec:
                if '+openmp' in spec:
                    fftflavor, fftlibs = 'fftw3-threads', '-lfftw3_omp -lfftw3 -lfftw3f'
                else:
                    fftflavor, fftlibs = 'fftw3', '-lfftw3 -lfftw3f'
                options.extend([
                    'FFTW3_CPPFLAGS={0}'.format(spec['fftw'].headers.cpp_flags),
                    'FFTW3_LIBS=-L{0} {1}'.format(
                        spec['fftw'].prefix.lib, fftlibs),
                ])
```

So far all of the tests I ran succeeded.

But if the FFTW3 interface is the only one officially supported with OpenMP, then we should switch to it when using mkl/intel-parallel-studio since MKL supports threading in there since MKL 11.1, especially since the package depends on the `fftw-api` already.

**Comment by glennpj** [comment]
It looks like I got the conflicts wrong in #23670. I was wrestling with `+openmp` and trying to use `fftw-api` and wound up in an unintended state. Looking at the `if...` clause, I had intended for `+openmp` to only be valid when `fftw` was the provider. I must have been playing around with mkl threading models along the way and did not set things back. In any event, I think the conflicts need to be set so that `+openmp` and `^mkl` are blocked because, based on comments in that PR, MKL FFTW is not thread safe. This diff would do it:
```
diff --git a/var/spack/repos/builtin/packages/abinit/package.py b/var/spack/repos/builtin/packages/abinit/package.py
index da16608..08c5552 100644
--- a/var/spack/repos/builtin/packages/abinit/package.py
+++ b/var/spack/repos/builtin/packages/abinit/package.py
@@ -96,18 +96,18 @@ class Abinit(AutotoolsPackage):
     # need openmp threading for abinit+openmp
     # TODO: The logic here can be reversed with the new concretizer. Instead of
     # using `conflicts`, `depends_on` could be used instead.
-    mkl_message = 'Need to set dependent variant to threads=openmp'
+    mkl_message = 'MKL FFTW is not compatible with +openmp'
     conflicts('+openmp',
               when='^fftw~openmp',
               msg='Need to request fftw +openmp')
     conflicts('+openmp',
-              when='^intel-mkl threads=none',
+              when='^intel-mkl',
               msg=mkl_message)
     conflicts('+openmp',
-              when='^intel-mkl threads=tbb',
+              when='^intel-oneapi-mkl',
               msg=mkl_message)
     conflicts('+openmp',
-              when='^intel-parallel-studio +mkl threads=none',
+              when='^intel-parallel-studio +mkl',
               msg=mkl_message)

     patch('rm_march_settings.patch', when='@:8')
```

**Comment by glennpj** [comment]
@gmatteo Can you take a look at this PR? You had commented about compatibility issues between MKL/FFTW for abinit in https://github.com/spack/spack/pull/23670#discussion_r647028515. I just reread that again and am wondering about the case where `intel-*-mkl` is the `blas`/`lapack` provider and `fftw` is the `fftw-api` provider. Is it the case that if `mkl` is the `blas`/`lapack` provider that it must also be the `fftw-api` provider?

**Comment by glennpj** [comment]
Well, I guess it is not actually possible to have mkl and fftw mixed anyway. That leads to:
```
==> Error: Multiple providers found for 'fftw-api': ['fftw', 'intel-mkl ^cpio']
```
if you try to force it. Otherwise, spack will use `mkl` as the `fftw-api` provider if `mkl` is in the spec, regardless of the setting for `fftw-api` in `packages.yaml`. I guess the only question is, can `mkl` be used with `+openmp`.

@dev-zero Which mkl thread model did you use with your build?

**Comment by dev-zero** [comment]
@glennpj I've been using `threads=openmp` (for `intel-parallel-studio+mkl`).
Wrt thread-safety of the FFTW3 implementation in MKL: it is thread-safe as of MKL 11.1, we have verified that thoroughly in CP2K, see https://github.com/cp2k/cp2k/blob/6471ac722e2ed247b997888a7786e72ad499f62f/src/pw/fft/fftw3_lib.F#L93-L106. But, it seems that when MKL is being used, Abinit does not even use its FFTW3 compatible API (selected by the fft flavor), but is instead using the MKL DFTI interface (fft flavor `dfti`), which seems to be OpenMP-safe and Abinit also spawns threads for DFTI (see `src/52_fft_mpi_noabirule/m_dfti.F90`).
Hence I don't think that those conflicts are needed, only that the part which pulls in the correct libraries for the `+openmp ^mkl` case needs to be fixed (see my latest commit I pushed).

**Comment by dev-zero** [comment]
Wrt mixing: that should be very well possible, but you may have to use the `providers` configuration in a `packages.yaml` (we have mixed MKL and FFTW3 before MKL 11 in CP2K). The problem is to properly figure out the provider of `fftw-api` inside the package (at the moment most packages rely on simply looking at whether `mkl` is in the spec or not).

File: var/spack/repos/builtin/packages/abinit/package.py [hunk 1]
```diff
@@ -209,27 +215,22 @@ def configure_args(self):
...
+                fftflavor = 'dfti'
+                options.extend([
+                    'FFT_CPPFLAGS={0}'.format(spec['fftw-api'].headers.cpp_flags),
+                    'FFT_LIBs={0}'.format(spec['fftw-api'].libs.ld_flags),
+                ])
+            elif '^fftw' in spec:
+                if '+openmp' in spec:
+                    fftflavor, fftlibs = 'fftw3-threads', '-lfftw3_omp -lfftw3 -lfftw3f'
+                else:
+                    fftflavor, fftlibs = 'fftw3', '-lfftw3 -lfftw3f'
```
Comment (glennpj) [inline var/spack/repos/builtin/packages/abinit/package.py]: A similar change will be needed for the `@:8` case.

**Review by glennpj — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/abinit/package.py (hunk 1 above)
Comment (dev-zero) [reply var/spack/repos/builtin/packages/abinit/package.py]: done

**Review by glennpj — APPROVED** [review]

**Pull request merged**